Add blog post for FSGroup on mount (#37621)
kubernetes/website: content/en/blog/_posts/2022-12-06-fsgroup-on-mount.md

---
layout: blog
title: "Kubernetes 1.26: Providing pod's FSGroup to CSI drivers during mount"
date: 2022-12-06
slug: kubernetes-12-06-fsgroup-on-mount
---

**Author:** Fabio Bertinatto (Red Hat), Hemant Kumar (Red Hat)

> We are happy to announce that the delegation of fsGroup to CSI Drivers feature has graduated to General Availability (GA) in Kubernetes 1.26.

Reviewer: Can you add when this feature was introduced as alpha, moved to beta?  Are there any blogs from previous releases related to this feature that can be referenced here?

Consequently, the CSI driver should apply the fsGroup to the files in the volume using a mount option. As an example, Azure File CSI Driver utilizes the `gid` mount option to map the fsGroup information to all the files in the volume.

It should be noted that in the example above the Kubelet will refrain from applying the permission changes into the volumes' files. In addition to that, `CSIDriver.spec.fsGroupPolicy` and `Pod.spec.securityContext.fsGroupChangePolicy` fields will have no effect.

## Why is it important?

Without this feature, applying the fsGroup information to files is not possible in certain storage environments.

For instance, Azure File does not support changing ownership and permissions of files. The CSI driver is only able to set the file permissions once, when it mounts the volume.

## How do I use it?

> This feature should be mostly transparent to users. If you maintain a CSI driver that should support this feature, head over to the [CSI Developer Documentation on FSGroup](https://kubernetes-csi.github.io/docs/support-fsgroup.html) for more information on how to support this feature in your CSI driver.

Reviewer: Any additional documentation for people to learn more, i.e., a link to the KEP?

> Existing CSI drivers that do not support this feature will continue to work as usual, i.e., they will not receive any fsGroup information from the Kubelet. In addition to that, the Kubelet will continue to perform the ownership and permissions changes to files according to the policies specified in `CSIDriver.spec.fsGroupPolicy` and `Pod.spec.securityContext.fsGroupChangePolicy`.

Reviewer: Can you give an example of a CSI driver already support this feature, i.e., Azure File CSI Driver?

Reviewer: That should be fine

---
layout: blog
title: "Kubernetes 1.26: Providing pod's FSGroup to CSI drivers during mount"

> date: 2022-12-06

Reviewer 2: ```suggestion
date: 2022-12-23
```

Please update the path `content/en/blog/_posts/2022-12-06-fsgroup-on-mount.md` to match.

slug: kubernetes-12-06-fsgroup-on-mount
---

> **Author:** Fabio Bertinatto (Red Hat), Hemant Kumar (Red Hat)

Reviewer 2: ```suggestion
**Authors:** Fabio Bertinatto (Red Hat), Hemant Kumar (Red Hat)
```

> The delegation of fsGroup to CSI Drivers feature was first introduced as alpha in Kubernetes 1.22, reaching the beta mark in Kubernetes 1.25. Today we are happy to announce that this feature has graduated to General Availability (GA) in Kubernetes 1.26.

Reviewer 2: ```suggestion
Delegation of `fsGroup` to CSI drivers was first introduced as alpha in Kubernetes 1.22,
and graduated to beta in Kubernetes 1.25.
For Kubernetes 1.26, we are happy to announce that this feature has graduated to
General Availability (GA). 
```

> Currently, if a fsGroup is specified in the [pod's security context](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod), all processes in the pod's containers will be part of the additional group specified in that field.

Reviewer 2: ```suggestion
In this release, if you specify a `fsGroup` in the
[security context](/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod),
for a (Linux) Pod, all processes in the pod's containers are part of the additional group
that you specified.
```

> In previous Kubernetes releases, the Kubelet would *always* apply the fsGroup ownership and permission changes to files in the volume according to the policy specified in the `Pod.spec.securityContext.fsGroupChangePolicy` field.

Reviewer 2: ```suggestion
In previous Kubernetes releases, the kubelet would *always* apply the
`fsGroup` ownership and permission changes to files in the volume according to the policy
you specified in the Pod's `.spec.securityContext.fsGroupChangePolicy` field.
```

> Starting with Kubernetes 1.26, CSI drivers have the option to apply the fsGroup settings during mount time of the volumes, freeing the Kubelet from changing the permissions of volumes.

Reviewer 2: ```suggestion
Starting with Kubernetes 1.26, CSI drivers have the option to apply the `fsGroup` settings during 
volume mount time, which frees the kubelet from changing the permissions of files and directories
in those volumes.
```

> Starting with Kubernetes 1.26, CSI drivers have the option to apply the fsGroup settings during attach or mount time of the volumes, freeing the Kubelet from changing the permissions of volumes.

Reviewer 3: nit: we should drop "during attach" from the sentence. I don't see how this can be applied during attach.

Author: dropped

## How does it work?

> CSI drivers that support this feature should advertise the [VOLUME_MOUNT_GROUP](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetcapabilities) node capability.

Reviewer 2: ```suggestion
CSI drivers that support this feature should advertise the
[`VOLUME_MOUNT_GROUP`](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetcapabilities) node capability.
```

> Initially, CSI drivers that support this feature should advertise the [VOLUME_MOUNT_GROUP](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetcapabilities) node capability.

Reviewer 3: Why - "Initially" ? For feature to work, the capability has to be present always.

Author: As a first step, at first

> Once recognizing this information, the Kubelet will pass the fsGroup information to the CSI driver during the pod startup time. This is done through the [NodeStageVolumeRequest](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodestagevolume) and [NodePublishVolumeRequest](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodepublishvolume) CSI calls.

Reviewer 2: ```suggestion
After recognizing this information, the kubelet will passes the `fsGroup` information to
the CSI driver during pod startup. This is done through the
[`NodeStageVolumeRequest`](https://github.com/container-storage-interface/spec/blob/v1.7.0/spec.md#nodestagevolume) and
[`NodePublishVolumeRequest`](https://github.com/container-storage-interface/spec/blob/v1.7.0/spec.md#nodepublishvolume)
CSI calls.
```
(notice that I pinned to CSI v1.7.0)

Author: `kubelet will passes`?

Author: I'll fix later

> Consequently, the CSI driver should apply the fsGroup to the files in the volume using a mount option. As an example, Azure File CSI Driver utilizes the `gid` mount option to map the fsGroup information to all the files in the volume.

Reviewer 2: ```suggestion
Consequently, the CSI driver is expected to apply the `fsGroup` to the files in the volume using a
_mount option_. As an example, Azure File CSIDriver utilizes the `gid` mount option to map
the `fsGroup` information to all the files in the volume.
```

> Consequently, the CSI driver should apply the fsGroup to the files in the volume using a mount option. As an example, Azure File CSI Driver utilizes the `gid` mount option to map the fsGroup information to all the files in the volume.

Reviewer: Add a link to Azure File CSI Driver github repo?
[Azure File CSI Driver](https://github.com/kubernetes-sigs/azurefile-csi-driver)

Author: Added.

> It should be noted that in the example above the Kubelet will refrain from applying the permission changes into the volumes' files. In addition to that, `CSIDriver.spec.fsGroupPolicy` and `Pod.spec.securityContext.fsGroupChangePolicy` fields will have no effect.

Reviewer 2: ```suggestion
It should be noted that in the example above the kubelet refrains from directly
applying the permission changes into the files and directories in that volume files. 
Additionally, two policy definitions no longer have an effect: neither
`.spec.fsGroupPolicy` for the CSIDriver object, nor
`.spec.securityContext.fsGroupChangePolicy` for the Pod.
```

> For more details about the inner workings of this feature, check out the [feature enhancement](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2317-fsgroup-on-mount/README.md) and the [CSI Developer Documentation](https://kubernetes-csi.github.io/docs/support-fsgroup.html).

Reviewer 2: ```suggestion
For more details about the inner workings of this feature, check out the
[enhancement proposal](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2317-fsgroup-on-mount/)
and the [CSI Driver `fsGroup` Support](https://kubernetes-csi.github.io/docs/support-fsgroup.html)
in the CSI developer documentation.
```

## Why is it important?

Without this feature, applying the fsGroup information to files is not possible in certain storage environments.

> For instance, Azure File does not support changing ownership and permissions of files. The CSI driver is only able to set the file permissions once, when it mounts the volume.

Reviewer 2: ```suggestion
For instance, Azure File does not support a concept of POSIX-style ownership and permissions
of files. The CSI driver is only able to set the file permissions at the volume level.
```

## How do I use it?

> This feature should be mostly transparent to users. If you maintain a CSI driver that should support this feature, head over to the [CSI Developer Documentation on FSGroup](https://kubernetes-csi.github.io/docs/support-fsgroup.html) for more information on how to support this feature in your CSI driver.

Reviewer 2: ```suggestion
This feature should be mostly transparent to users. If you maintain a CSI driver that should
support this feature, read
[CSI Driver `fsGroup` Support](https://kubernetes-csi.github.io/docs/support-fsgroup.html)
for more information on how to support this feature in your CSI driver.
```

> Existing CSI drivers that do not support this feature will continue to work as usual, i.e., they will not receive any fsGroup information from the Kubelet. In addition to that, the Kubelet will continue to perform the ownership and permissions changes to files according to the policies specified in `CSIDriver.spec.fsGroupPolicy` and `Pod.spec.securityContext.fsGroupChangePolicy`.

Reviewer 2: ```suggestion
Existing CSI drivers that do not support this feature will continue to work as usual:
they will not receive any `fsGroup` information from the kubelet. In addition to that,
the kubelet will continue to perform the ownership and permissions changes to files
for those volumes, according to the policies specified in `.spec.fsGroupPolicy` for the
CSIDriver and `.spec.securityContext.fsGroupChangePolicy` for the relevant Pod.
```
